Dockerfile

```diff
@@ -0,0 +1,54 @@
+# Use the official Alpine-based Rust image as a parent image
+FROM rust:1.80-alpine AS builder
+
+# Set the working directory in the container
+WORKDIR /usr/src/app
+
+# Install build dependencies
+RUN apk add --no-cache \
+    musl-dev \
+    openssl-dev \
+    openssl-libs-static \
+    pkgconfig \
+    patch
+
+# Set environment variables for static linking
+ENV OPENSSL_STATIC=yes
+ENV OPENSSL_DIR=/usr
+
+# Copy the current directory contents into the container
+COPY . .
+
+# Set up build arguments for architecture detection
+ARG TARGETARCH
+
+# Set the Rust target based on the detected architecture
+RUN case "$TARGETARCH" in \
+        "amd64")  echo "x86_64-unknown-linux-musl" > /tmp/target ;; \
+        "arm64")  echo "aarch64-unknown-linux-musl" > /tmp/target ;; \
+        *)        echo "Unsupported architecture: $TARGETARCH" && exit 1 ;; \
+    esac
+
+# Add the target to rustup and build the application
+RUN RUST_TARGET=$(cat /tmp/target) && \
+    rustup target add $RUST_TARGET && \
```

Comment:
Didn't know rustup could do that
Trying to rework the Dockerfile now

Reply:
You don't really need `target` commands unless you're trying to cross-compile

Reply:
Tried simplifying the following section (as of 6a07a2f47cffe27c2a2b5d50d8097d65e701ae59):
```
# Determine the target architecture and build the application
RUN RUST_TARGET=$(rustc -vV | sed -n 's/host: //p') && \
    rustup target add $RUST_TARGET && \
    RUSTFLAGS='-C target-feature=+crt-static' cargo build --release --target $RUST_TARGET
```
to just:
```
RUN RUSTFLAGS='-C target-feature=+crt-static' cargo build --release
```
but the change doesn't seem to play well with `docker buildx`?

As an example, see the following error when running `docker buildx build --platform linux/amd64,linux/arm64 -t perennial131/inv_sig_helper:latest -f Dockerfile --push .`:
```
5.679 error: cannot produce proc-macro for `futures-macro v0.3.30` as the target `aarch64-unknown-linux-musl` does not support these crate types
------
Dockerfile:23
--------------------
  21 |     
  22 |     
  23 | >>> RUN RUSTFLAGS='-C target-feature=+crt-static' cargo build --release
  24 |     
  25 |     # Create a new stage for a smaller final image
--------------------
ERROR: failed to solve: process "/bin/sh -c RUSTFLAGS='-C target-feature=+crt-static' cargo build --release" did not complete successfully: exit code: 101
```

Reply:
Apparently `proc_macro` is only a dynamic library, it cannot be linked statically

Reply:
> Apparently `proc_macro` is only a dynamic library, it cannot be linked statically [rust-lang/rust#40174](https://github.com/rust-lang/rust/issues/40174)

Skimming the linked issue, seems like it **is** possible but the build needs to be run on ["something like Ubuntu, Fedora, or another full system"](https://github.com/rust-lang/rust/issues/40174#issuecomment-538791091), meaning you [don't get binaries precompiled for musl](https://github.com/rust-lang/rust/issues/40174#issuecomment-569785489) when you build from a system like Alpine

Reply:
You mean a system that uses glibc? Maybe that should be used for the builder system

Reply:
Revisited using a Debian-based Rust image in the build stage and it seems to be working in ff9a37856463062f31af2d1aa00a87dc23bedb6c

However, compiling a static binary via `RUN RUSTFLAGS='-C target-feature=+crt-static' cargo build --release` doesn't seem possible without using the `cargo --target` flag then hard-coding architectures in the Dockerfile since buildx's TARGETARCH doesn't directly map to Rust target triples (see the output of `rustc --print target-list` for reference)

Reply:
Last ditch effort: install glibc on Alpine or a distribution that uses such, Using musl with Rust is a bit of a pain in the ass for the time being

Reply:
> Last ditch effort: install glibc on Alpine or a distribution that uses such, Using musl with Rust is a bit of a pain in the ass for the time being. It doesn't help that the Rust compiler works best with glibc

Hence the switch to Debian-based images in ff9a37856463062f31af2d1aa00a87dc23bedb6c
At the moment I'm still trying to figure out static binaries, but honestly I don't think it's worth the added complexity in the build process

Reply:
> > Last ditch effort: install glibc on Alpine or a distribution that uses such, Using musl with Rust is a bit of a pain in the ass for the time being. It doesn't help that the Rust compiler works best with glibc
> 
> Hence the switch to Debian-based images in [ff9a378](https://github.com/iv-org/inv_sig_helper/commit/ff9a37856463062f31af2d1aa00a87dc23bedb6c); no need to install toolchains or such At the moment I'm still trying to figure out static binaries, but honestly I don't think it's worth the added complexity in the build process

You can install glibc in Alpine, with `apk add gcompat libstdc++`

Reply:
> You can install glibc in Alpine, with `apk add gcompat libstdc++`

Any particular reason for doing this versus just using a Debian-based image?

Reply:
> > You can install glibc in Alpine, with `apk add gcompat libstdc++`
> 
> Any particular reason for doing this versus just using a Debian-based image?

Alpine is still much lighter than Debian, other than that, not much difference I think?

Reply:
> Alpine is still much lighter than Debian, other than that, not much difference I think?

Size of the image only really matters for the final stage since that's whats actually being deployed in Docker

Reply:
I'm fine with Debian too. Alpine is not always the best: https://www.theregister.com/2023/05/16/alpine_linux_318/

Reply:
Just tested on ARM64 host and seems like the binary file produced didn't work:
```
# docker run -it sig_helper
exec /app/inv_sig_helper_rust: no such file or directory
```

Reply:
This men explains very well that he had to generate the binary from outside of Docker then copy it to the Docker image: https://bxbrenden.github.io/. I'm not very found because it means that a simple `docker build` won't work.

@perennialtech What are you trying to achieve here? You want to build a multi arch binary directly from an x86 host? What about letting the QEMU virtualization do its job? It's going to be slower, but still works anyway.
Let the binary being built on each host architecture.

Reply:
@unixfox The [latest revision of the Dockerfile](https://github.com/iv-org/inv_sig_helper/blob/ca4e3f8d832b9df8b397a25076b55f00211fff58/Dockerfile) just uses `RUN cargo build --release`, but I don't think the ARM64 binary actually works since it's not statically linked when built from a glibc system (mentioned in the article you linked)

The arch detection via rustup was to template in the correct value for the `cargo build --target` flag in an attempt to solve [errors like this](https://github.com/iv-org/inv_sig_helper/pull/13#discussion_r1715328985) when trying to rely solely on `buildx` + QEMU

Reply:
I think it's ok to do it via rustup for the time being. If it's the only choice.